ciri/consensus/snapshot/local_snapshots/pruning_service.c

```diff
@@ -0,0 +1,335 @@
+
+static retcode_t collect_transactions_for_pruning_do_func(flex_trit_t *hash, iota_stor_pack_t *pack, void *data,
+                                                          bool *should_branch, bool *should_stop);
+
+static uint64_t get_last_snapshot_to_prune_index(pruning_service_t *const ps);
+
+static void *pruning_service_routine(void *arg);
+
+typedef struct collect_transactions_for_pruning_do_func_params_s {
+  uint64_t current_snapshot_index;
+  tangle_t const *const tangle;
+  hash243_set_t *const transactions_to_prune;
+} collect_transactions_for_pruning_do_func_params_t;
+
+static retcode_t collect_unconfirmed_future_transactions_for_pruning_do_func(flex_trit_t *hash, iota_stor_pack_t *pack,
+                                                                             void *data, bool *should_branch,
+                                                                             bool *should_stop) {
+  retcode_t ret = RC_OK;
+  *should_stop = false;
+  *should_branch = false;
+
+  if (pack->num_loaded == 0) {
+    return RC_OK;
+  }
+
+  if (transaction_snapshot_index((iota_transaction_t *)pack->models[0]) == 0) {
+    *should_branch = true;
+  }
+
+  return RC_OK;
+}
+
+static retcode_t collect_transactions_for_pruning_do_func(flex_trit_t *hash, iota_stor_pack_t *pack, void *data,
+                                                          bool *should_branch, bool *should_stop) {
+  retcode_t ret = RC_OK;
+  *should_stop = false;
+  *should_branch = false;
+  collect_transactions_for_pruning_do_func_params_t *params = data;
+
+  if (pack->num_loaded == 0) {
+    return RC_OK;
+  }
+
+  uint64_t current_snapshot_index = transaction_snapshot_index((iota_transaction_t *)pack->models[0]);
+
+  if (current_snapshot_index >= params->current_snapshot_index) {
+    *should_branch = true;
+    ERR_BIND_RETURN(
+        tangle_traversal_dfs_to_future(params->tangle, collect_unconfirmed_future_transactions_for_pruning_do_func,
+                                       hash, params->transactions_to_prune, &params),
+        ret);
+  }
+
+  if (ret) {
+    *should_stop = true;
+    return ret;
+  }
+
+  return RC_OK;
+}
+
+static void *pruning_service_routine(void *arg) {
+  pruning_service_t *ps = (pruning_service_t *)arg;
+  tangle_t tangle;
+  uint64_t start_timestamp, end_timestamp;
+  uint64_t start_index;
+  spent_addresses_provider_t sap;
+  DECLARE_PACK_SINGLE_MILESTONE(milestone, milestone_ptr, milestone_pack);
+
+  {
+    connection_config_t db_conf = {.db_path = ps->conf->tangle_db_path};
+
+    if (iota_tangle_init(&tangle, &db_conf) != RC_OK) {
+      log_critical(logger_id, "Failed in initializing db\n");
+      return NULL;
+    }
+  }
+
+  {
+    connection_config_t db_conf = {.db_path = ps->spent_addresses_service->conf->spent_addresses_db_path};
+
+    if (ps->spent_addresses_service->conf->spent_addresses_files != NULL) {
+      if (iota_spent_addresses_provider_init(&sap, &db_conf) != RC_OK) {
```

Comment:
You always need to initialize the provider, it doesn't depends on the files. The files are for the spent addresses service, not provider.